documents important MC vars and procs and adds tracking of tick usage when the mc first wakes up and other stats

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
![tick diagram2](https://user-images.githubusercontent.com/15794172/145672748-7f3871a6-c046-4f2d-8142-d4573bbb929c.png)
this was part of an effort to further understand the MC and help quantify how much time it has to actually run within the average tick. other sleeping procs can resume before the MC does which pushes it out of the tick and we dont seem to track that, so i added that with average_starting_tick_usage which is a running average the tick usage of when the MC starts running for the tick. 

![Screenshot_1668](https://user-images.githubusercontent.com/15794172/146275960-da16eecd-ef79-4581-ad16-cb56e5c8d4fe.png)
this is what the current new Master entry looks like. It was split up into Master and Tick Division entries.

- TickRate has been renamed to Ticks Per Iteration since thats a less ambiguous name for the processing var
- Max Tick Limit Im still not sure what to do with since im 99% sure the one you see in the stat entry is whatever the running tick limit was when SSstat_panels was fire()'d which means it doesnt have a meaning relevant to master as a whole
- Sleeping Overtime im not sure if ill keep but it is how much sleeping procs resumed after the MC start after the scheduled end of the tick that master fired in.
- stoplag Threads is how many concurrent stoplag()'s there are - might be useful in showing tick contention on live i intend to experiment with it (note from live testing: debatably useful)
- Ticks Skipped is a running average of how many ticks the MC goes to sleep for minus 1 (because sleeping for one tick is normal, it just makes it resume on the next tick). this is a stat otherwise completely invisible to the stat panel but leads to entire missed ticks of potential processing time so i think this one is pretty good

The Tick Division entry shows measurements of what each portion of the tick took up. Meant to show whats eating up the average tick beyond maptick and the MC itself.
- Pre MC is a running average of the tick usage when master's main loop resumes. If this is >0 it means there are sleeping procs resuming before master which can cause master to not process if it starts too late
- MC shows the average tick usage of each master controller run (ending tick usage minus starting tick usage)
- Post MC is how much tick usage sleeping procs take after the MC has gone to sleep but before SendMaps processes. as predicted by mso's old comments this is much less common than pre MC usage because most sleeps tend to be longer than the MC's sleeps and sleeps are ran in the order theyre called.
- Maptick has been moved into this category (and renamed from Internal Tick Rate)
- Post Maptick is an attempt to measure the tick usage of client verbs/procs executing after SendMaps processes. this one is likely the least accurate since 1. the measurements are only updated for verbs/procs that have a define that sets a tick usage var to the ending tick usage so not all of them actually update it and 2. it has to use world.map_cpu to determine that portions tick usage, which is an average of the last x ticks instead of measuring a single tick. However I think it could be a useful measurement we'll see


![Screenshot_1651](https://user-images.githubusercontent.com/15794172/145748920-4f1baa5b-42ad-4e87-968e-0bd00d1d9869.png)
also added an additional stat on every subsystems stat_entry which is the percentage of the average tick this subsystem takes up whenever it ignites, added at the end of the default stat_entry message. Its essentially a running average of time spent running / time spent not running. So in this example lighting takes 0.1% of all recent processing and machines takes 1%. This is important because it makes it easier to see what time a subsystem takes from the server as opposed to most other stats which show hwo long a given subsystem takes to complete its arbitrary run of processing

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
In general, works to quantify and log more information about what goes on within a tick.

also documents and clarifies some things about the MC
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
documents important MC vars and procs and adds tracking of starting usage

Kylerace committed:
changes the average to MC_AVG_FAST_UP_SLOW_DOWN to emphasize outliers

Kylerace committed:
so long sweet prince

LemonInTheDark commented on code/controllers/master.dm:
```diff
@@ -342,6 +352,9 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 		tickdrift = max(0, MC_AVERAGE_FAST(tickdrift, (((REALTIMEOFDAY - init_timeofday) - (world.time - init_time)) / world.tick_lag)))
 		var/starting_tick_usage = TICK_USAGE
+
+		average_starting_tick_usage = MC_AVG_FAST_UP_SLOW_DOWN(average_starting_tick_usage, starting_tick_usage)
```
could you log this in with td and maptick? I want to graph

Kylerace replied:
how could i forget

LemonInTheDark replied:
thx xoxo

Kylerace committed:
makes the average starting tick usage logged

MrsTonedOne committed:
Track sleep resumes that happen after the mc

this code is verified to be untested

Kylerace committed:
Merge pull request #6 from MrStonedOne/patch-576

MrStonedOne commented on code/controllers/master.dm:
```diff
@@ -55,13 +68,19 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
+	///!Are we loading in a new map?
+	var/map_loading = FALSE
+
+	///!for scheduling different subsystems for different stages of the round
```
if you're gonna move these to before the thing, you have to remove the `!` from the comment as that tells codedocs to attach the comment to the *previous* thing, not the next thing.

Kylerace replied:
ah right

Kylerace replied:
fixed

MrStonedOne commented on code/controllers/master.dm:
```diff
@@ -289,34 +308,33 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 	var/list/tickersubsystems = list()
 	var/list/runlevel_sorted_subsystems = list(list()) //ensure we always have at least one runlevel
 	var/timer = world.time
-	for (var/thing in subsystems)
-		var/datum/controller/subsystem/SS = thing
-		if (SS.flags & SS_NO_FIRE)
+	for (var/datum/controller/subsystem/subsystem as anything in subsystems)
+		if (subsystem.flags & SS_NO_FIRE)
 			continue
-		SS.queued_time = 0
-		SS.queue_next = null
-		SS.queue_prev = null
-		SS.state = SS_IDLE
-		if (SS.flags & SS_TICKER)
-			tickersubsystems += SS
+		subsystem.queued_time = 0
+		subsystem.queue_next = null
```
You have been docked 420 social credit score points.

Kylerace replied:
![image](https://user-images.githubusercontent.com/15794172/145674214-a74cbbb9-237b-476e-b7fc-c5a46790ac9b.png)

Kylerace committed:
*shakes fist* the weed man strikes again

MrsTonedOne reviewed (commented):
@github-actions github-actions bot requested a review from MrStonedOne 3 minutes ago

Kylerace committed:
adds logging for the other two measurements and fixes docs

LemonInTheDark commented on code/controllers/master.dm:
```diff
@@ -615,24 +657,23 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 
 /datum/controller/master/stat_entry(msg)
-	msg = "(TickRate:[Master.processing]) (Iteration:[Master.iteration]) (TickLimit: [round(Master.current_ticklimit, 0.1)])"
+	msg = "(TickRate:[Master.processing]) (Iteration:[Master.iteration]) (TickLimit: [round(Master.current_ticklimit, 0.1)]) (Average Starting Tick Usage: [round(Master.average_starting_tick_usage, 0.1)] \
+	(Average Sleeping TIck Usage: [round(Master.average_sleeping_tick_usage, 0.1)]) (Average Sleeping Overtime: [round(Master.average_sleeping_overtime_usage)]))"
```
Do we need the average here? more similar words there are the harder it is to read at a glance

Kylerace replied:
i shortened it to Avg ill see if that makes it better as i adjust it more (and add more stats)

LemonInTheDark commented on code/controllers/subsystem/time_track.dm:
```diff
@@ -59,6 +59,9 @@ SUBSYSTEM_DEF(time_track)
 			"tidi_avg",
 			"tidi_slowavg",
 			"maptick",
+			"average starting tick usage",
+			"average sleeping tick usage",
+			"average sleeping tick overtime",
```
_ seperated please, also avg_start_tick_use maybe? or just sleep_use_avg and sleep_over_avg? less longwinded the better, while still remaining clear and all

Kylerace committed:
fixes stat collection for lemon

Kylerace committed:
master code improvement, adds tracking for stoplag's & MC skipped ticks

LemonInTheDark commented on code/__DEFINES/MC.dm:
```diff
@@ -79,3 +79,14 @@
 }\
 /datum/controller/subsystem/processing/##X
+
+///RunQueue() didnt complete a run and something fucked up
+#define RUNQUEUE_ERRORED FALSE
+///RunQueue() successfully completed a run
+#define RUNQUEUE_SUCCESSFUL_RUN TRUE
+
+///CheckQueue() wasnt able to go through its list of subsystems and something fucked up
+#define CHECKQUEUE_ERRORED FALSE
+///CheckQueue() was able to go through its list of subsystems
+#define CHECKQUEUE_SUCCESSFUL_RUN TRUE
```
Why not just TRUE/FALSE?

Kylerace replied:
eh probably better to keep it to TRUE/FALSE i just thought they were important enough to warrant the define

LemonInTheDark replied:
It's the same pattern as most other procs that succeed/fail right?

Kylerace committed:
the reported ticklimit is wrong either way but its less wrong with this

Kylerace committed:
adds tracking of average cost per tick for each subsystem

Kylerace committed:
adds docs to the atmos markdown for average_percentage_of_tick

Kylerace committed:
adds more stats to time_track and minor cleanup in master and lighting

Kylerace committed:
minor docs

Kylerace commented:
average_percentage_of_tick is inaccurate in some way, i suspect its when a subsystem frequently pauses during a fire and resumes the next tick

Kylerace committed:
makes statpanels/fire() have some god damn newlines

Kylerace committed:
adds support for tracking post-maptick tick usage

Kylerace committed:
fixes average_percentage_of_tick (mostly) and docs

Kylerace committed:
splits up the Master tab into Master and Tick Division

Tick Division shows what portions of the tick (pre MC sleeps, the MC itself, post MC sleeps, maptick, and post maptick client verbs/procs) takes up individually

Kylerace committed:
removes maptick from BYOND statpanel entry since its in another one now

Kylerace committed:
whoops mixed spacing

Kylerace commented:
doing other stuff ill reopen this later

LemonInTheDark commented:
you had better
